module/zfs/vdev_queue.c

```diff
@@ -174,6 +174,7 @@ int zfs_vdev_async_write_active_max_dirty_percent = 60;
  * they aren't able to help us aggregate at this level.
  */
 int zfs_vdev_aggregation_limit = 1 << 20;
+int zfs_vdev_aggregation_limit_non_rotating = SPA_OLD_MAXBLOCKSIZE;
```

Comment:
I wonder if for non-rotational devices it might make sense to set this even lower.  I think it does make sense to have different limits for rotational and non-rotations devices.  Regarding the `1MB` size for the normal `zfs_vdev_aggregation_limit` as I recall this was something of a mixed bag.  You can make a legitimate case for both the higher (1M) or lower (128K) values depending on exactly what hardware and workload you're targeting.

Related to this there is a pending bit of work for the ABD code to allow us to eliminate this copy under certain circumstances by sharing the same pages between ABDs.  But thus far it hasn't been a priority.

Reply:
I was just trying to be safe in case vdev while not rotating, have other problems with short writes (misaligned CoW, etc), but indeed I could see sense in reducing this to about 32-64KB for average SSDs.  At I/Os smaller then 32KB time spent on other code parts of ZFS is much higher to make this less important.

If somebody have good thoughts what would be better here, please comment.

Reply:
We usually start with no aggregation for SSDs, but that causes problems for hybrid pools. If we get some time, we can try to test with various options for SSDs, but for the enterprise class SSDs we normally use, it is unlikely to make much difference.

Reply:
If someone can spare the time, I think it would be useful to do a survey of both of the aggregation limits in order to settle on good defaults for most systems.